### PR TITLE
 Porting schedules (except convolutions) to C++

### PR DESCRIPTION
**Pull request by alex-weaver**
## Operators
- [x] broadcast_min, broadcast_max
- [x] elementwise: identity, negative, log, left_shift, right_shift, clip, cast
- [x] reduction: sum, max, min, argmax, argmin
- [x] transforms: expand_dims, transpose, reshape, squeeze, concatenate, split
- [x] batch_norm_inference
- [x] binary_dense
- [x] dense
- [x] dilate
- [x] leaky_relu
- [x] flatten
- [x] scale_shift_nchw, scale_shift_nhwc
- [x] pad - add pad_value argument
- [x] global_pool, pool
- [x] softmax, log_softmax
- [x] pow, broadcast_pow
## Generic schedules
- [x] injective
- [x] _default_schedule
- [x] extern

## CUDA schedules
- [x] injective
- [x] dense
- [x] extern
- [x] pool, global_pool
- [x] reduce
- [x] softmax

## Other schedules
- [x] x86: binarize_pack, binary_dense
- [x] rocm: dense
- [x] x86: injective, default from #775

### CODE REVIEW

**Comment by joyalbin** [comment]
How about "power" operator in elementwise category?

**Comment by alex-weaver** [comment]
power would definitely be useful. Would it make sense to define two?
1. pow(Tensor, Expr)
2. pow(Tensor, Tensor) - where this one follows broadcasting rules

**Comment by joyalbin** [comment]
yes, two versions of pow make sense.
I am working on implementation of basic version first, will post the progress here later.

**Comment by alex-weaver** [comment]
@tqchen ok that should be all of the non-conv ops and schedules for #749 
let me know if anything needs fixing re reviewing.

**Comment by alex-weaver** [comment]
@joyalbin I've added an implementation for both pow functions to this PR. The key thing it needed was adding the "pow" intrinsic to ir_operator.h like the other unary functions.

**Comment by alex-weaver** [comment]
ok the merged branch fails, but it's failing on an unrelated test. The actual changes are in commit 
14c7cdf

**Comment by tqchen** [comment]
that testcase can be fixed by retriggering, which is fine. Please add unit-testcases to the functions added, I will do a round of review today. Thanks for the effort!

File: topi/include/topi/broadcast.h [hunk 1]
```diff
@@ -147,6 +147,67 @@ inline tvm::Tensor broadcast_mod(const tvm::Tensor& A,
...
+*
+* \param A The first tensor
+* \param B The second tensor
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor whose op member is a pointwise maximum with broadcast
+*/
+inline tvm::Tensor broadcast_max(const tvm::Tensor& A,
+  const tvm::Tensor& B,
```
Comment (tqchen) [inline topi/include/topi/broadcast.h]: code style, align argument to previous bracket.`


```c++
broadcast_max(const tvm::Tensor& A,
                         const tvm::Tensor& B)
```

File: topi/include/topi/elemwise.h [hunk 2]
```diff
@@ -28,6 +28,150 @@ TOPI_DECLARE_UNARY_OP(exp);
...
+* \brief Creates an operation that performs pointwise left shift by n bits
+*
+* \param x The input tensor
+* \param n The number of bits to shift by
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor whose op member is the left shift operation
+*/
+inline Tensor left_shift(const Tensor& x,
```
Comment (tqchen) [inline topi/include/topi/elemwise.h]: change this to operator overloading. 

```c++
inline Tensor operator<<(const Tensor &x, const Expr&x);
```

File: topi/include/topi/nn.h [hunk 3]
```diff
@@ -86,6 +115,7 @@ inline tvm::Tensor relu(const tvm::Tensor& t,
 inline tvm::Tensor pad(const tvm::Tensor& t,
                        const tvm::Array<tvm::Expr>& pad_before,
                        tvm::Array<tvm::Expr> pad_after = tvm::Array<tvm::Expr>(),
+                       Expr* pad_value = nullptr,
```
Comment (tqchen) [inline topi/include/topi/nn.h]: We can simply allow pass in pad_value as ```Expr()``` and use pad_value.defined() to see if it is defined

File: topi/include/topi/detail/pad_utils.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+#define TOPI_DETAIL_PAD_UTILS_H_
+
+#include <vector>
+
+#include "tvm/tvm.h"
+
+namespace topi {
+using namespace tvm;
+
+std::vector<int> GetPadTuple(int pad_h, int pad_w) {
```
Comment (tqchen) [inline topi/include/topi/detail/pad_utils.h]: add a short comment on what detail  util functions do

File: topi/include/topi/schedules/x86/bnn.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
+/*!
+*  Copyright (c) 2017 by Contributors
+* \file x86/bnn.h
+* \brief x86 schedule for binary operations
+*/
+#ifndef TOPI_SCHEDULES_X86_BNN_H_
+#define TOPI_SCHEDULES_X86_BNN_H_
+
+#include "topi/tags.h"
+#include "topi/detail/fuse.h"
```
Comment (tqchen) [inline topi/include/topi/schedules/x86/bnn.h]: use same namespace rules as python. Put schedules under x86, generic, cuda. Remove schedules folder.

**Comment by tqchen** [comment]
@sxjscience can you do a correctness review on the broadcast and reduce logics of this PR thanks!

**Comment by alex-weaver** [comment]
Thanks for the feedback, I'll get those fixed. A couple of things:
1. Since we don't have numpy to serve as a reference in unit tests, how should this be handled? Is it sufficient to use simple fixed inputs to validate against hand-calculated outputs?
2. I wasn't sure whether some of the util functions (mainly those in detail/fuse.h and detail/extern.h) belong in TVM core rather than TOPI. Let me know if these should be moved

**Comment by tqchen** [comment]
We can expose things to python for the test, which we have to do eventually. See https://github.com/dmlc/tvm/tree/master/src/api for some examples. Let us register them as
The general steps are:
- [ ] Add a topi/src/topi.cc to register APIs
- [ ] Register functions under ```topi.xx``` 
- [ ] Add this as TVM extension which allows us to test correctness from python side  
     - Refer to https://github.com/dmlc/tvm/tree/master/apps/extension

**Comment by tqchen** [comment]
It might be helpful to do some API coverage test in cpp, but I will let you be the judge

**Comment by sxjscience** [comment]
@tqchen Is it very urgent? I can review the code this Friday.

**Comment by tqchen** [comment]
I think it is fine as you as you review it before the end of the week

**Comment by alex-weaver** [comment]
@tqchen I'm just looking at exposing the ops to python, how should the register code be built? should tvm.cc be added to the build for the main tvm lib, or should it build as a separate library?

**Comment by tqchen** [comment]
For now we can build it as a separate library and load it in via ctypes.LoadLibrary https://github.com/dmlc/tvm/tree/master/apps/extension is a good reference

File: topi/python/topi/_ffi/libinfo.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+"""Library information."""
+from __future__ import absolute_import
```
Comment (tqchen) [inline topi/python/topi/_ffi/libinfo.py]: it is likely we don't need these ffi utilities, but can simply reuse all things under TVM.
The find lib path could be an exception but it is also fine if we give a search_path.

Comment (tqchen) [reply topi/python/topi/_ffi/libinfo.py]: if there is a need to enhance tvm's ffi utilities, let us do so

**Comment by alex-weaver** [comment]
@tqchen I'm not sure why this isn't passing the build - from the log it is building libtopi.so without error, but it's not being packed/unpacked. The PR contains a modification to the Jenkinsfile - will the CI system be using the one in this commit or the latest merged one? If it's using the last merged, what is the best way to resolve this? could the Jenkinsfile be modified in a separate PR without breaking the build if libtopi.so is missing?

File: topi/python/topi/_ffi/base.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        A char pointer that can be passed to C API
+    """
+    return ctypes.c_char_p(string.encode('utf-8'))
+
+
+def c_array(ctype, values):
+    """Create ctypes array from a python array
+
+    Parameters
+    ----------
```
Comment (tqchen) [inline topi/python/topi/_ffi/base.py]: this file can be safely removed

File: topi/include/topi/cuda/reduction.h [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+*
+* \return A schedule for the given ops.
+*/
+Schedule schedule_reduce(const Target& target, Array<Tensor> outs) {
+  Array<Operation> out_ops;
+  for (auto t : outs) {
+    out_ops.push_back(t->op);
+  }
+  auto s = create_schedule(out_ops);
+  TraverseAfterReduce(target, s, outs[0]->op);
```
Comment (sxjscience) [inline topi/include/topi/cuda/reduction.h]: Do we need to assert that the length of `outs` is equal to 1?

File: topi/python/topi/_ffi/cpp_init.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
+"""FFI implementation details"""
+from __future__ import absolute_import as _abs
+import sys
+from tvm._ffi.function import list_global_func_names, get_global_func, _get_api
+
+def _init_api(module_name, prefix):
```
Comment (tqchen) [inline topi/python/topi/_ffi/cpp_init.py]: can also reuse tvm python for this, fix tvm python if necessary

File: topi/python/setup.py [hunk 10]
```diff
@@ -13,7 +14,40 @@
     from setuptools import setup
     from setuptools.extension import Extension
 
-__version__ = "0.1.0"
+def get_lib_names():
+    if sys.platform.startswith('win32'):
+        return ['libtopi.dll', 'topi.dll']
```
Comment (tqchen) [inline topi/python/setup.py]: change name to libtvm_topi

File: topi/python/topi/cpp.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
+"""FFI for C++ TOPI ops and schedules"""
+from ._ffi.cpp_init import _init_api
+
```
Comment (tqchen) [inline topi/python/topi/cpp.py]: it is likely we only need to optionally load libtopi if it can be found and run the cpp init. Let us not require libtvm_topi for now, and just use it for test purpose. Once logic transfer to c++ complete, we can mark it as required

Comment (alex-weaver) [reply topi/python/topi/cpp.py]: Ok so you're meaning add tests, but don't add them to the CI build yet?

Comment (alex-weaver) [reply topi/python/topi/cpp.py]: And should I reset the jenkinsfile to the current master?

Comment (tqchen) [reply topi/python/topi/cpp.py]: we do need to add them to topi tests. Just not require user to do this. I am not sure what is happening on the Jenkinsfile side and why things didn't get packed

**Comment by tqchen** [comment]
Note there are some warnings in compilation that is better to be fixed http://mode-gpu.cs.washington.edu:8080/blue/organizations/jenkins/dmlc%2Ftvm/detail/PR-763/46/pipeline/36#step-74-log-330

File: topi/python/topi/cpp.py (hunk 11 above)
Comment (tqchen) [reply topi/python/topi/cpp.py]: As a temporary solution, let us build libtopi_tvm at the cpp test phase and do the test there

Comment (alex-weaver) [reply topi/python/topi/cpp.py]: ok, would that be running the python test of the exported ops/schedules in the cpp test phase?

File: topi/include/topi/reduction.h [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+* \param keepdims If this is set to true, the axes which are reduced are
+* left in the result as dimensions with size one. This enables the result
+* to broadcast correctly against the input array.
+*
+* \return The result tensor.
+*/
+Tensor CommReduceIdx(const Tensor& data,
+                     const std::vector<int>& axis,
+                     FCommReduce func,
+                     bool keepdims = false) {
```
Comment (sxjscience) [inline topi/include/topi/reduction.h]: Is it possible to merge `CommReduceIdx` and `CommReduce`? Most of the codes are the same.

**Comment by sxjscience** [comment]
I've added some minor comments. The overall logic looks good to me and is similar to the python implementation.

File: topi/include/topi/reduction.h (hunk 12 above)
Comment (alex-weaver) [reply topi/include/topi/reduction.h]: They are very similar, but annoyingly different in some control flow. An approach I was thinking about was pulling shared code out into separate functions - eg. functions to calculate reduce_axes and target_shape. That would cut down much of the duplication. Would that be ok?

File: topi/python/topi/cpp.py (hunk 11 above)
Comment (tqchen) [reply topi/python/topi/cpp.py]: yes

File: topi/include/topi/reduction.h (hunk 12 above)
Comment (sxjscience) [reply topi/include/topi/reduction.h]: Yes, that should be okay.

Comment (tqchen) [reply topi/include/topi/reduction.h]: any update wrt this comment?

File: topi/include/topi/nn/dilate.h [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+* dimensions and any layout.
+* \param strides Dilation stride for each dimension. Stride 1
+* means no dilation.
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return The output tensor.
+*/
+inline Tensor dilate(const Tensor& x,
+                     std::vector<int> strides,
```
Comment (tqchen) [inline topi/include/topi/nn/dilate.h]: Change most of the std::vector<int> arguments to Array<Expr>

Comment (alex-weaver) [reply topi/include/topi/nn/dilate.h]: How should an array of ints be represented? Should it be an Array<Expr> filled with constants? (Array<int> fails to compile)

Comment (alex-weaver) [reply topi/include/topi/nn/dilate.h]: re: /reduction.h (I can't reply directly for some reason), I've pulled out most of the duplication into 2 common functions: MakeReduceAxes and MakeReduceTargetShape they should be in the latest commit

File: topi/include/topi/reduction.h (hunk 12 above)
Comment (alex-weaver) [reply topi/include/topi/reduction.h]: I've pulled most of the duplication out into 2 common functions: MakeReduceAxes and MakeReduceTargetShape, they should be in the latest commit.

File: topi/include/topi/nn/dilate.h (hunk 13 above)
Comment (tqchen) [reply topi/include/topi/nn/dilate.h]: Array of <Expr> note that strides can be passed in as symbolic expressions, although not frequently used

Comment (alex-weaver) [reply topi/include/topi/nn/dilate.h]: Sorry array of what? Github may have broken your comment.

I see that strides could be expressions, but nn/dilate.h:69 relies on being able to evaluate strides[i]. This and surrounding lines are:
```
        if (strides[i] != 1) {
          index_tuple.push_back(indices[i] / strides[i]);
          not_zero.push_back((indices[i] % strides[i]) == 0);
        } else {
          index_tuple.push_back(indices[i]);
        }
```

Would a correct implementation of this with changing the signature to `Array<Expr> strides` be this?
```
        if (IsConstInt(strides[i]) && GetConstInt(strides[i]) != 1) {
          index_tuple.push_back(indices[i] / strides[i]);
          not_zero.push_back((indices[i] % strides[i]) == 0);
        } else {
          index_tuple.push_back(indices[i]);
        }
```

Comment (tqchen) [reply topi/include/topi/nn/dilate.h]: yes, let us do that

File: topi/include/topi/detail/extern.h [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  auto elem_offset = Expr();
+  return BufferNode::make(data, dtype, shape, Array<Expr>(), elem_offset, name, "",
+    -1, 0);
+}
+
+/*! \brief A function wrapping the invocation of an external function */
+using FExtern = std::function<Expr(Array<Buffer>, Array<Buffer>)>;
+
+/*! \brief Create tensors representing the result of invoking an external function */
+Array<Tensor> make_extern(const Array<Array<Expr>>& out_shapes,
```
Comment (tqchen) [inline topi/include/topi/detail/extern.h]: ```>> -> > >``` note the space, for backward compatibility for MSVC

File: topi/include/topi/detail/extern.h [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+  Array<Tensor> outputs;
+  for (size_t i = 0; i < output_placeholders.size(); ++i) {
+    outputs.push_back(op.output(i));
+  }
+  return outputs;
+}
+
+/*! \brief Pack a buffer object to be used as an argument to a PackedFunc */
+Expr pack_buffer(Buffer buf) {
```
Comment (tqchen) [inline topi/include/topi/detail/extern.h]: this belongs to a detail namespace. Add more comments to argument and return value.

This functions is used to create a DLTensor structure on heap to be able to pass a symbolic buffer as arguments to TVM PackedFunc

File: topi/include/topi/detail/fuse.h [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+*/
+#ifndef TOPI_DETAIL_FUSE_H_
+#define TOPI_DETAIL_FUSE_H_
+
+#include "tvm/tvm.h"
+
+namespace topi {
+using namespace tvm;
+
+/*! \brief Fuse all of the given args */
```
Comment (tqchen) [inline topi/include/topi/detail/fuse.h]: document all arguments

File: topi/include/topi/detail/pad_utils.h [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+#define TOPI_DETAIL_PAD_UTILS_H_
+
+#include <vector>
+
+#include "tvm/tvm.h"
+
+namespace topi {
+using namespace tvm;
+
+/*! \brief Get padding size for each side given padding height and width */
```
Comment (tqchen) [inline topi/include/topi/detail/pad_utils.h]: document all arguments, and what is expected in return

File: topi/include/topi/cuda/reduction.h (hunk 8 above)
Comment (tqchen) [reply topi/include/topi/cuda/reduction.h]: We will need to add the check here

**Comment by tqchen** [comment]
Thanks for all the effort to bring this set of changes! I like the structure overall and the code looks good as I skim through. 

Since there are a lot of changes. Please take a pass over the code to make sure we have good readability of the code, put in comments delibrately in places that you think requires explaination so it is easier for others to understand and improve upon this.

**Comment by alex-weaver** [comment]
Ok I've fixed up the comments on some of the more obscure bits of the code. Let me know if there's anything else that needs sorting. I'd like to get some kind of schedule registry working like there is in the python code but I think that's best left for a separate PR ;)

**Comment by tqchen** [comment]
Thanks! this is merged
